### ISSUE TITLE
Build with USE_DISTRIBUTED=0 fails

### ISSUE THREAD

**Issue by nairbv** [pytorch/pytorch#36870]
## Bug

In linux on a devgpu, when trying to build with USE_DISTRIBUTED=0 I get:
`nccl.h: No such file or directory`

I think nccl should only be used if USE_DISTRIBUTED=1, so maybe we have an incorrect `#ifdef` somewhere. Changing USE_DISTRIBUTED=1 gets my build to succeed.

## Environment

$ python ~/bin/collect_env.py
Collecting environment information...
PyTorch version: 1.6.0a0+d5f531d
Is debug build: Yes
CUDA used to build PyTorch: 9.2

OS: CentOS Linux 7 (Core)
GCC version: (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
CMake version: version 3.14.0

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 9.2.88
GPU models and configuration:
GPU 0: Tesla M40
GPU 1: Tesla M40

Nvidia driver version: 396.69
cuDNN version: /usr/local/cuda-9.2/targets/x86_64-linux/lib/libcudnn.so.7.4.2

Versions of relevant libraries:
[pip] numpy==1.17.2
[pip] numpydoc==0.9.1
[pip] torch==1.6.0a0+de0cd40
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.4                      243
[conda] mkl-include               2020.0                      166
[conda] mkl-service               2.3.0            py37he904b0f_0
[conda] mkl_fft                   1.0.14           py37ha843d7b_0
[conda] mkl_random                1.1.0            py37hd6b4f25_0
[conda] torch                     1.6.0a0+140ac1f          pypi_0    pypi

## Additional context

I started seeing this after rebasing a PR on master (I usually leave distributed off to make builds faster), so probably related to a recent change. Maybe https://github.com/pytorch/pytorch/pull/36193 ?
cc @ezyang @gchanan @zou3519 @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528 @osalpekar @malfet

**Comment by albanD** [comment]
Triage review to make sure this is high priority

**Comment by mrshenli** [comment]
Do we know if this occurs for 1.5 release as well?

**Comment by mrshenli** [comment]
Seems so, trying a fix.

```
In file included from misc/nvmlwrap.cc:7:0:                                                                                                                          │·························
include/nvmlwrap.h:129:1: error: ‘ncclResult_t’ does not name a type; did you mean ‘nvmlReturn_t’?                                                                   │·························
 ncclResult_t wrapNvmlSymbols(void);                                                                                                                                 │·························
 ^~~~~~~~~~~~                      
```

**Comment by ezyang** [comment]
It sounds like we also need to add a CI config to test `USE_DISTRIBUTED=0`, if we consider bugs like this high priority

**Comment by mrshenli** [comment]
Looks like the `python_nccl.cpp` was added to support [`torch/cuda/nccl.py`](https://github.com/pytorch/pytorch/blob/release/1.5/torch/cuda/nccl.py), but the latter is a legacy API added > 2 years ago, which are not tested or documented. I am not aware if anyone is using this. If not, we probably should delete that stack, including `python_nccl.cpp`. Add #36940 to track.

**Comment by mrshenli** [comment]
first attempt: #36947

**Comment by mrshenli** [comment]
I verified that this is indeed a regression. This error didn't occur on 1.4 release branch.

**Comment by mrshenli** [comment]
The code that links NCCL to `TORCH_PYTHON_LINK_LIBRARIES` was moved to `caffe2/CMakeLists.txt` from `torch/CMakeLists.txt` in  #27124. The new location of this link was later removed by #30315 to split the dependencies into `torch_hip` and `torch_cuda`, which seems missed to adding it back into `TORCH_PYTHON_LINK_LIBRARIES`.

**Comment by mrshenli** [comment]
closed by #36948
